transport/src/main/java/io/netty5/channel/AbstractChannel.java

```diff
@@ -1942,28 +1921,75 @@ public Buffer allocateBuffer() {
             return readBufferAllocator.allocate(readBufferAllocator(), readHandle.estimatedBufferCapacity());
         }
 
-        void complete(ReadHandleFactory.ReadHandle readHandle) {
+        private void complete() {
+            try {
+                readSomething();
+            } finally {
+                continueReading = false;
+                readLoopComplete();
+            }
+        }
+
+        private boolean completeFailure(Throwable cause) {
+            try {
+                readSomething();
+
+                pipeline().fireChannelExceptionCaught(cause);
+
+                if (cause instanceof PortUnreachableException) {
+                    return false;
+                }
+                // If oom will close the read event, release connection.
+                // See https://github.com/netty/netty/issues/10434
+                return cause instanceof IOException && !(AbstractChannel.this instanceof ServerChannel);
+            } finally {
+                continueReading = false;
+                readLoopComplete();
+            }
+        }
+
+        private void readSomething() {
             // Check if something was read as in this case we wall need to call the *ReadComplete methods.
             if (readSomething) {
                 readSomething = false;
-                readHandle.readComplete();
                 pipeline().fireChannelReadComplete();
             }
+            readHandle.readComplete();
         }
 
-        boolean completeFailure(ReadHandleFactory.ReadHandle readHandle, Throwable cause) {
-            complete(readHandle);
-            pipeline().fireChannelExceptionCaught(cause);
-
-            if (cause instanceof PortUnreachableException) {
-                return false;
+        void readLoop() {
+            boolean closed;
+            try {
```

Comment:
Perhaps set the `continueReading` to an initial state, in case `doReadNow` doesn't call back to `processRead`.